### PR TITLE
Create RequestManager and SubscriptionManager classes

### LINKED ISSUE #4429: Create `RequestManager` class to handle internal web3 requests
### Description

We need to create `RequestManager` class to handle existing internal request handling. This package would migrate all the logic for existing `web3-core-requestmanager` package from `1.x`.

https://github.com/ChainSafe/web3.js/blob/50afcd0318112b5da12f4944d3fdbdbb2f862ca2/packages/web3-core-requestmanager/src/index.js#L40-L46

### Acceptance Criteria

- The class should cover existing logic available in `1.x` package 
- The class should be exported from `web3-core` package
- Any utility structure for that class should not be exported from the package. 

### Additional Information

- The class should not have any dependency on available providers, wherever necessary use `BaseProvider`

### LINKED ISSUE #4430: Create `SubscriptionManager` class to handle internal web3 requests
### Description

We need to create `SubscriptionManager` class to handle existing internal request handling. This package would migrate all the logic for existing `web3-core-subscriptions` package from `1.x`.

https://github.com/ChainSafe/web3.js/blob/50afcd0318112b5da12f4944d3fdbdbb2f862ca2/packages/web3-core-subscriptions/src/index.js#L28-L33

### Acceptance Criteria

- The class should cover existing logic available in `1.x` package 
- The class should be exported from `web3-core` package
- Any utility structure for that class should not be exported from the package. 

### Additional Information

- The class should not have any dependency on available providers, wherever necessary use `BaseProvider`
- The class should use `RequestManager` as dependency if required

### PR DESCRIPTION
**Pull request by nazarhussain**
## Description

Create the implementation of `RequestManager` and `SubscriptionManager`. 

Fixes #4429, #4430

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [x] I ran `npm run test:unit` with success.
- [x] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [x] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [x] I have tested my code on the live network.
- [x] I have checked the Deploy Preview and it looks correct.
- [x] I have updated the `CHANGELOG.md` file in the root folder.

### CODE REVIEW

File: packages/web3-common/src/types.ts [hunk 1]
```diff
@@ -29,9 +29,9 @@ export interface JsonRpcResponseWithResult<T = JsonRpcResult> {
 	readonly result: T;
 }
 
-export type JsonRpcResponse<T = JsonRpcResult> =
-	| JsonRpcResponseWithError
-	| JsonRpcResponseWithResult<T>;
+export type JsonRpcResponse<R = JsonRpcResult, E = JsonRpcResult> =
```
Comment (Spacesai1or) [inline packages/web3-common/src/types.ts:32]: We should formalize how we're going to specify these generics

Currently the rest of the repo uses `T, T2, etc.`, above you use:

```typescript
export class InvalidResponseError<ErrorType = unknown>
```

and now here you're using `R and E`.

My preference would to be explicit and use `ReturnType, ErrorType, etc.`

File: packages/web3-common/src/web3_base_provider.ts [hunk 2]
```diff
@@ -25,22 +19,21 @@ export const JSONRPC_ERR_UNSUPPORTED_METHOD = 4200;
...
+	public static isWeb3Provider(provider: unknown) {
+		return (
+			provider instanceof Web3BaseProvider ||
+			Boolean(provider && (provider as { [symbol]: boolean })[symbol])
+		);
+	}
+
+	// eslint-disable-next-line class-methods-use-this
+	public get [symbol]() {
+		return true;
```
Comment (Spacesai1or) [inline packages/web3-common/src/web3_base_provider.ts]: Can you add documentation for what's going on here?

File: packages/web3-core/src/web3_subscription_manager.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	): Promise<InstanceType<ST[T]>> {
+		if (!this.requestManager.provider) {
+			throw new InvalidProviderError('No provider set');
+		}
+
+		if (!this.supportsSubscriptions()) {
+			throw new Error('The current provider do not support subscriptions');
+		}
+
+		const Klass = this.registeredSubscriptions[name];
```
Comment (Spacesai1or) [inline packages/web3-core/src/web3_subscription_manager.ts:36]: `Klass`?

File: packages/web3-core/src/web3_request_manager.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+): provider is LegacySendAsyncProvider => 'sendAsync' in provider;
+
+export class Web3RequestManager extends Web3EventEmitter<{
+	[key in Web3RequestManagerEvent]: SupportedProviders;
+}> {
+	private _requestCounter = 0;
+	private _provider!: SupportedProviders;
+	private readonly _providers: { [key: string]: Web3BaseProviderConstructor };
+
+	public constructor(provider: Web3BaseProvider | string, net: Socket) {
```
Comment (Spacesai1or) [inline packages/web3-core/src/web3_request_manager.ts]: Should `net` be optional, or `Socket` could be an additional type for `provider`, so:

```typescript
public constructor(provider: Web3BaseProvider | string | Socket) {
```

File: packages/web3-core/src/web3_request_manager.ts [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+export const isLegacySendProvider = (
+	provider: SupportedProviders,
+): provider is LegacySendProvider => 'send' in provider;
+
+export const isLegacySendAsyncProvider = (
+	provider: SupportedProviders,
+): provider is LegacySendAsyncProvider => 'sendAsync' in provider;
+
+export class Web3RequestManager extends Web3EventEmitter<{
```
Comment (Spacesai1or) [inline packages/web3-core/src/web3_request_manager.ts:25]: This should have a public `request` method, no?

File: packages/web3-core/src/web3_request_manager.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+	[key in Web3RequestManagerEvent]: SupportedProviders;
```
Comment (Spacesai1or) [inline packages/web3-core/src/web3_request_manager.ts:26]: this resolves to type `any`

File: packages/web3-common/src/types.ts (hunk 1 above)
Comment (nazarhussain) [reply packages/web3-common/src/types.ts:32]: I earlier used `T1`, `T2` which is common in TS space. But later I observed with complex generics its better to use full names, so why latest code is using the full name and will be using similar onward.

File: packages/web3-core/src/web3_subscription_manager.ts (hunk 3 above)
Comment (nazarhussain) [reply packages/web3-core/src/web3_subscription_manager.ts:36]: It returns the `classes` not the the `objects`, so why used this name.

File: packages/web3-core/src/web3_request_manager.ts (hunk 4 above)
Comment (nazarhussain) [reply packages/web3-core/src/web3_request_manager.ts]: To use existing code logic made it optional.

File: packages/web3-core/src/web3_request_manager.ts (hunk 5 above)
Comment (nazarhussain) [reply packages/web3-core/src/web3_request_manager.ts:25]: For using existing name logic I used `send` if you prefer `request` we can rename it.

File: packages/web3-core/src/web3_request_manager.ts (hunk 6 above)
Comment (nazarhussain) [reply packages/web3-core/src/web3_request_manager.ts:26]: May be IDE is not loading up correctly. I tried in a test script and used tsc and it works fine.

File: packages/web3-common/src/types.ts (hunk 1 above)
Comment (Spacesai1or) [reply packages/web3-common/src/types.ts:32]: Okay, we need to update all other instances to use full name as well - created [an issue](https://github.com/ChainSafe/web3.js/issues/4499) to track this

File: packages/web3-core/src/web3_request_manager.ts [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+				throw new ProviderError(`Can't autodetect provider for "${provider}'"`);
+			}
+		}
+
+		this.emit(Web3RequestManagerEvent.BEFORE_PROVIDER_CHANGE, this.provider);
+		this._provider = newProvider ?? null;
+		this.emit(Web3RequestManagerEvent.PROVIDER_CHANGED, this.provider);
+	}
+
+	public async send<ResultType, RequestType = unknown>(request: JsonRpcRequest<RequestType>) {
```
Comment (Spacesai1or) [inline packages/web3-core/src/web3_request_manager.ts:85]: What about the [sendBatch](https://github.com/ChainSafe/web3.js/blob/1.x/packages/web3-core-requestmanager/src/index.js#L198) method?

Comment (nazarhussain) [reply packages/web3-core/src/web3_request_manager.ts:85]: We have a separate issue for this https://github.com/ChainSafe/web3.js/issues/4432

**Review by Spacesai1or — APPROVED** [review]

**Review by luu-alex — APPROVED** [review]

File: packages/web3-common/src/web3_event_emitter.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+export interface Web3Emitter<T extends Web3EventMap> {
+	on<K extends Web3EventKey<T>>(eventName: K, fn: Web3EventCallback<T[K]>): void;
+	off<K extends Web3EventKey<T>>(eventName: K, fn: Web3EventCallback<T[K]>): void;
+	emit<K extends Web3EventKey<T>>(eventName: K, params: T[K]): void;
+}
+
+export class Web3EventEmitter<T extends Web3EventMap> implements Web3Emitter<T> {
+	private readonly _emitter = new EventEmitter();
+
+	public on<K extends Web3EventKey<T>>(eventName: K, fn: Web3EventCallback<T[K]>) {
```
Comment (jdevcs) [inline packages/web3-common/src/web3_event_emitter.ts:16]: It should return Emitter for chained subscriptions. 

also there must be support of `once, removeListener, removeAllListeners`, ... etc 

IMO, there should be more control to user on EventEmitter for making it flexible, and we can do that by extending EventEmitter by Web3EventEmitter

File: packages/web3-core/src/utils.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+export const isLegacySendProvider = (
+	provider: SupportedProviders,
+): provider is LegacySendProvider => 'send' in provider;
+
+export const isLegacySendAsyncProvider = (
+	provider: SupportedProviders,
+): provider is LegacySendAsyncProvider => 'sendAsync' in provider;
+
+export const isSupportSubscriptions = (provider: SupportedProviders): boolean => {
```
Comment (jdevcs) [inline packages/web3-core/src/utils.ts:24]: `isSupportSubscriptions` should be enforced by baseProvider and implemented/handled in actual providers only. so I recommend to remove it from web3-core utils.

File: packages/web3-core/src/web3_config.ts [hunk 10]
```diff
@@ -14,13 +15,18 @@ interface ConfigOptions {
...
 
+export enum Web3ConfigEvent {
+	CONFIG_CHANGE = 'CONFIG_CHANGE',
+}
+
+export abstract class Web3Config
+	extends Web3EventEmitter<{ [Web3ConfigEvent.CONFIG_CHANGE]: ConfigEvent<ConfigOptions> }>
+	implements ConfigOptions
+{
 	private _config: ConfigOptions = {
```
Comment (jdevcs) [inline packages/web3-core/src/web3_config.ts:30]: should be named as eth1Config or `executionLayerConfig`

**Review by jdevcs — CHANGES_REQUESTED** [review]

File: packages/web3-core/src/web3_request_manager.ts [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+			} else if (/^ws(s)?:\/\//i.test(provider)) {
+				newProvider = new this.providers.WebsocketProvider(provider);
+
+				// IPC
+			} else if (typeof net === 'object' && typeof net.connect === 'function') {
+				newProvider = new this.providers.IpcProvider(provider, net);
+			} else {
+				throw new ProviderError(`Can't autodetect provider for "${provider}'"`);
+			}
+		}
```
Comment (jdevcs) [inline packages/web3-core/src/web3_request_manager.ts:78]: if `provider` is not string but injected provider it must also set that.

File: packages/web3-common/src/web3_event_emitter.ts (hunk 8 above)
Comment (nazarhussain) [reply packages/web3-common/src/web3_event_emitter.ts:16]: There should be lesser control for user and more for us. There should be more usability for the user in the provided API interfaces. 

e.g. For some of the use cases we don't want user to have `emit` on public interface (instance level) and we just want user to have event listener there. 

This above example provide the fact that we as library author need more control on what interfaces to make available for the users. So why we shouldn't extend from the `EventEmitter`, which will just expose unnecessary and sometime even vulnerabilities to the users.

File: packages/web3-core/src/utils.ts (hunk 9 above)
Comment (nazarhussain) [reply packages/web3-core/src/utils.ts:24]: This logic is coupled with the `RequestManager` so why is part of `web3-core`. The `Web3BaseProvider` already have the public abstract interface for it. The function `isSupportSubscriptions` is using that interface to check.

File: packages/web3-core/src/web3_config.ts (hunk 10 above)
Comment (nazarhussain) [reply packages/web3-core/src/web3_config.ts:30]: Its the the private variable, never been exposed to anywhere in the library, so no matter what name to use.

Comment (jdevcs) [reply packages/web3-core/src/web3_config.ts:30]: better readability

File: packages/web3-common/src/web3_event_emitter.ts (hunk 8 above)
Comment (jdevcs) [reply packages/web3-common/src/web3_event_emitter.ts:16]: how will I as user can listen **once** for event, or remove listeners ?

Comment (nazarhussain) [reply packages/web3-common/src/web3_event_emitter.ts:16]: For removing listener we have `off` interface and I just added the support for `once` as well.

File: packages/web3-core/src/web3_config.ts (hunk 10 above)
Comment (nazarhussain) [reply packages/web3-core/src/web3_config.ts:30]: I think as of now we have only this set of configuration so lest's just use `Web3Config` name as it is. Once we have some other set of configurations we can refactor.

**Review by jdevcs — APPROVED** [review]

**Pull request merged**